SQLAlchemy 2.0 upgrades to ORM usage in /lib

jdavcs opened the pull request:
## SA 2.0 ORM usage to-do:
- [x] lib/galaxy/actions
- [x] lib/galaxy/celery
- [ ] lib/galaxy/datatypes
- [ ] lib/galaxy/jobs
- [ ] lib/galaxy/managers
- [ ] lib/galaxy/metadata
- [ ] lib/galaxy/model
- [x] lib/galaxy/queue_worker
- [x] lib/galaxy/quota
- [x] lib/galaxy/security
- [ ] lib/galaxy/tool_shed
- [x] lib/galaxy/util
- [x] lib/galaxy/visualization
- [ ] lib/galaxy/web
- [x] lib/galaxy/webapps/base
- [x] lib/galaxy/webapps/galaxy/api
- [ ] lib/galaxy/webapps/galaxy/controllers
- [x] lib/galaxy/webapps/galaxy/services
- [x] lib/galaxy/webapps/reports (except enable_eagerloads)
- [x] lib/galaxy/workflow
- [ ] lib/galaxy_test
- [ ] lib/tool_shed


## Requires more work:
- [ ] lib/galaxy/job_execution (requires addressing SessionlessContext, which implements a small subset of SQLAlchemy session's API)
- [x] lib/galaxy/tools (same as above: SessionlessContext)
- [ ] lib/galaxy/authnz (requires rewriting all the authnz unit tests, because they are all based on a mocked subset of SQLAlchemy)
- [ ] lib/galaxy/webapps/reports: handle `.query.enable_eagerloads()`
- [x] Fix the `FormDefinition.to_dict` bug ([ref](https://app.element.io/#/room/#galaxyproject_backend:gitter.im/$TJCv12X_UdHUSYQXTvERHD1R7dTIZFQPPWbePf2MkpQ)) 

Ref: #12541

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs committed:
Fix SA2.0 ORM usage in webapps.galaxy.controllers.history

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.actions

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.celery

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.quota

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.security

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.visualization

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.queue_worker

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.webapps.base; refactor

Drop unused parameters and logic from select methods in legacy controller.

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.tools

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.workflow

jdavcs committed:
Fix SA2.0 ORM usage in galaxy.webapps/reports [partially]

jdavcs committed:
Fix SA2.0 ORM usage in in galaxy.jobs [partially]

Move data access method into manager (get_jobs_to_check_at_startup)

mvdbeek commented on lib/galaxy/model/repositories/__init__.py:
```diff
@@ -0,0 +1,36 @@
+MappedType = Base
+
+
+class BaseRepository:
```
What is this solving ? It looks like a lot of overhead

jdavcs replied:
You mean the class `BaseRepository`? It's just a base class for all repositories which contains common functionality. So, instead of having `self.session.get(self.Foo, primary_key`) in 100+ repository classes, you have it in one.

mvdbeek replied:
Can you elaborate some more on this ? I may not get it, but for instance
`parent_folder = trans.sa_session.query(trans.app.model.LibraryFolder).get(parent_id)` that you're replacing with `parent_folder = LibraryFolderRepository(trans.sa_session).get(parent_id)` does not look more efficient and I don't think is something you'd do if you're familiar with sqlalchemy (unless that's a documented new pattern ?).

It seems that this would also be problematic if you're doing further queries and mainpulations where you're expecting ORM objects.

jdavcs replied:
Sure! `session.query(Foo).get(pkey)` is SQLAlchemy-specific code. It's low level and it's tightly coupled with SA's API. When that API changes (and it did), or when we decide to use a different API (for whatever reason - like filter vs where, scalars vs. execute, etc.), we'll have to change that particular chain of method calls in every place we use it - there are many hundreds, if not thousands of those, which are also not easily grepped ("get" is so generic). With a repository, you change it in one place only. There's that.

Then there's code comprehension: whereas the `session.get(id)` vs. `foo_repository.get(id)` is not particularly convincing, any other statement is: `session.scalars(select(Foo).limit(1)).first()` to return the first Foo is less readable than `foo_repository.get_first()`. 

If one has to write low-level SA code to do any kind of data access, that makes learning low-level SA a prerequisite for writing any code that interacts with the db. Besides, being familiar with SQLAlchemy doesn't necessarily guarantee that, for example, one would know to use `limit(1)` in the case above, or remember to do that. In other words, in my opinion, wrapping library-level data access code in a repository abstraction makes the code both more maintainable (by far!) and more comprehensible.

That said, you are absolutely right about the challenge of more complex queries and manipulations: we certainly don't want to rewrite parts of SA's API! My hope is that we'll be able to tuck away the vast majority of such low level code into repositories, and eliminate a ton of duplication in the process. But yes, it's possible that there will be some logic that is too complex to be abstracted away without adding a lot of conceptual overhead - if that's the case, we'll leave such code as is, of course. I'm working my way up from trivial to simple (those are duplicated all over the code base) - to more complex, so by the time we have complex db logic to untangle, there'll be a ton of improvement done already.

Data access repositories is a pattern. I'll provide a more detailed description as it relates to our code base in the PR description. Essentially, it boils down to eliminating duplication and encapsulating low level code/logic in objects and methods with names that describe *what* is being done/returned (e.g. foo repo returning all deleted bars) as opposed to *how* it is implemented (e.g. session/scalars/select/first/etc...)

mvdbeek replied:
> tuck away the vast majority of such low level code into repositories

that IMO is not the right direction. I am -0.9 on this change, sorry.

mvdbeek replied:
> Data access repositories is a pattern.

More complicated stuff like those you've mentioned belongs into Managers, which I think is the right, and currently used, abstraction for this.

jdavcs replied:
I'll try to provide a better justification in the PR description, I hope the pattern can win you over :) Managers, I think, are there to handle the business logic, not the low-level details of database access.

jmchilton replied:
I'm... persuadable in either direction on introducing this pattern... but I will say the tool parameter sql alchemy usage is a good example of why the pattern might be useful. I like the change of moving low-level logic out of there and I don't think managers would be a great fit because I think they belong above that level of the dependency chain.

My concern with adding the abstraction layer would be... are we making simple things a little more complex and not helping complex things at all? I think a lot of the base manager stuff is problematic in this way - a deep layer of complexity that only works for 3 object types and just makes the rest of the code feel inconsistent.

mvdbeek replied:
I'm sorry, I don't think you will be able to write something to convince me while you have code like this:
```python
class QuotaRepository(BaseRepository):
    def __init__(self, session: SessionType):
        super().__init__(session, Quota)

    def get(self, primary_key: int) -> Quota:
        return cast(Quota, super().get(primary_key))
```

Anything is better than that. If this is about providing `get_first` and other helpers in a convenient way you could subclass or compose a different session type.

mvdbeek replied:
I know it's still draft and so not entirely fair to draw conclusions, but

> def get(self, primary_key: int) -> Quota:

looks wrong to me, since get returns `Model | None`. That leads me to worry about insulating ourselves from proper types

jdavcs replied:
It's not ready yet - there have been many iterations: I'm changing it almost with every commit - adapting, simplifying as I factor out more code from other modules. The class above can be reduced to 
```py
class QuotaRepository(BaseRepository):
    def __init__(self, session: SessionType):
        super().__init__(session, Quota)
```
The `get` method override is there just to provide a `Quota` object as the return value, not a `MappedType` - that's all. That gives you much tighter type checking (which has already revealed previously hidden bugs: #16553, #16550). Remove all those and all those `FooRepository.get(pkey)` will still work - that's where the `BaseRepository` comes in. The `__init__` method is there to facilitate other quota-related data access methods: it sets the `Quota` as the `self.model_class`.

@jmchilton your concern is spot on, but I'm sure we are covered. Here are just a few examples:

| SQLAlchemy                                    | using repos        |
| --------------------------------------------- | -------------------|
| session.get(Foo, 5)                           | foo_repo.get(5)    |
| session.scalars(select(Foo)).all()            | foo_repo.get_all() |
| session.execute(select(Foo)).scalar_one()     | foo_repo.one()     |
| session.scalars(select(Foo).limit(1)).first() | foo_repo.first()   | 
| session.scalar(select(func.count(Foo.id)))    | foo_repo.count()   | 

Again, some logic may be too complex to factor out - I don't know yet. But some can be easily wrapped in a method with a descriptive name, like [here](https://github.com/galaxyproject/galaxy/pull/16434/files#diff-5503de21527ed99e95c7c848f727292eb4e04c606bc1ec0b9695cf3bccb3913bL42). There are plenty more vivid examples - the same SQLAlchemy code repeated over and over again, with the only difference being the model object - I just haven't gotten to implementing those yet.

jdavcs replied:
> > def get(self, primary_key: int) -> Quota:
> 
> looks wrong to me, since get returns `Model | None`. That leads me to worry about insulating ourselves from proper types

That's not the intent, of course. The correct return type should be `Optional[Quota]` (so here's one more iteration).

mvdbeek replied:
> That gives you much tighter type checking

to me these are highly questionable redundant annotations, given that everything returned through the session already has the proper type:

```
# (variable) users1: Sequence[User]
users1 = session.scalars(select(User)).all()

# (variable) user: User
user = session.query(User).one()

# (variable) user_iter: Iterator[User]
user_iter = iter(session.scalars(select(User)))
```
see https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#sql-expression-typing-examples



> session.execute(select(Foo)).scalar_one()	foo_repo.one()

that's not always going to be what we want for something called `.one()`, so why not roll with the upstream syntax ? It's not like the sqlalchemy people hate good syntax / UX.

jdavcs replied:
> to me these are highly questionable redundant annotations, given that everything returned through the session already has the proper type:

Yes, very nice indeed - it appears SA already does the right thing! Which means we can simplify repository code and get rid of that redundant typing!

> that's not always going to be what we want for something called `.one()`, so why not roll with the upstream syntax ? It's not like the sqlalchemy people hate good syntax / UX.

I'll have the data to justify this (or not justify it) shortly - I'm not there yet. It's been my impression so far that there's a sufficient amount of repetitive and fairly detailed SA code that can be easily factored out into a repo class without hiding proper types and without the need to invent any new syntax, etc.

The main goal of this PR is updating the SA code in our code base to SA 2.0. The repos came as a natural addition, after I had copied and pasted the same SA code a sufficient number of times. However, the PR is still primarily about updating the SA code. The repos are, actually, making the process easier (having `FooRepository.whatever` is like placing a check mark next to the fixed code block). Most importantly, this is not an irreversible change: replacing FooRepo.get_whatever with the corresponding SA code would be a trivial sed exercise. So, in the end, if it appears the layer isn't justified, we can just remove it - the PR will still update the SA code to 2.0.

jmchilton replied:
> Here are just a few examples:

These are not complicated examples IMO - these are exactly the simple code.

jmchilton replied:
Something like ``get_with_filter_order_by_hid`` - is indeed code that I think belongs in the model layer and should be tested in isolation. I think though just a static method on the HDA class or a function in that module that took in the sa_session along with the rest of the arguments you've created and just did the work directly would be simpler and feel more "python" to me.

jdavcs replied:
Yes, you're right - those are quite simple. I'll move on to more complex ones - I hope I'll find better examples.

The `get_with_filter_order_by_hid` - bad approach altogether: I'm doing exactly what we shouldn't do - taking perfectly normal SA syntax and restating the same imperative logic as an ugly method name which doesn't improve anything beyond reducing duplicate code. Just an iteration - so please ignore. 

I think I have a better option - just pushed. The method name, I think, should say *what* is being fetched/done. The repo module (or some module within the model) contains all the low level SA code that details *how* it's done.  See [`repository.hda`](https://github.com/galaxyproject/galaxy/pull/16434/files#diff-fbf8c38da6c2f286bd70d2fb60c2178981f10b765ad66566d0baa247bb53bc16R15) and [`webapps.galaxy.services.history.py`](https://github.com/galaxyproject/galaxy/pull/16434/files#diff-13b6a7251d924008c6cfe1de85cbabc93dc77f12f32c98a059914843b82fcd64R644). Naturally, there will be a lot of similar methods with different names - that can be then refactored in the repositories to eliminate any duplicate constructs.

jdavcs committed:
Fix SA2.0 ORM usage in webapps.galaxy.services

Move data access method to managers.quotas (get_quotas)
Move data access method to managers.histories (get_len_files_by_history)
Move data access method to managers.groups (get_group_by_name)

jdavcs committed:
Fix SA2.0 ORM usage in managers.api_keys; refactor

Replace python iteration with SA batch update

jdavcs committed:
Fix SA2.0 ORM usage in managers.cloud

jdavcs committed:
Fix SA2.0 ORM usage in managers.dbkeys

jdavcs committed:
Fix SA2.0 ORM usage in managers.group_roles

jdavcs committed:
Fix SA2.0 ORM usage in managers.group_users

jdavcs committed:
Fix SA2.0 ORM usage in managers.groups

Move data access method to managers.users

jdavcs committed:
Fix SA2.0 ORM usage in webapps.galaxy.api

Move data access method to managers.workflows (count_stored_workflow_user_assocs)
Move data access method to managers.pages (get_page_revisions)

jdavcs committed:
Fix SA2.0 ORM usage in galaxy/tool_shed/util

jdavcs committed:
Refactor get_user_by_username, get_user_by_email; use across code base

mvdbeek commented on lib/galaxy/webapps/galaxy/services/histories.py:
```diff
@@ -640,11 +641,7 @@ def get_custom_builds_metadata(
-            .filter_by(history=history, extension="fasta", deleted=False)
-            .order_by(model.HistoryDatasetAssociation.hid.desc())
-        )
+        fasta_hdas = get_fasta_hdas_by_history(trans.sa_session, history.id)
```
seems easier to read inline and isn't used anywhere else, so maybe just keep it here ? otherwise bind it to the History class ?

jdavcs replied:
I think it would be better to move low-level SA code out of services and into a dedicated location (class, module, package - whatever). In my opinion, it's easy to read it inline when one has read it many times, or has read such code many times; otherwise one has to switch conceptual context and go from business logic to "what exactly does this retrieve from the database". Having a clearly named method/function instead eliminates this step (get HDAs by history), thus reducing cognitive load. One reads code faster that way, I think.

As for `model.History`: I think it would be great if we could limit the model classes to model definitions only + definitions and functionality related to model instances. This function is fetching related data from the database - I think we should place it elsewhere. Since we haven't agreed yet on what to do with the repositories idea, I think putting them into appropriate manager modules as lightweight standalone functions (also, as per @jmchilton's suggestion) is a reasonable compromise?

mvdbeek replied:
>  I think though just a static method on the HDA class or a function in that module that took in the sa_session along with the rest of the arguments you've created and just did the work directly would be simpler and feel more "python" to me.

?

mvdbeek replied:
this is never going to be used anywhere else, so having it at least in this module is an upside, since it's got nothing to do with the manager

jdavcs replied:
Done.

Yes, John meant the model class, I was wrong about that. However, I do think this is low-level code that belongs in a separate location, not all over the code base - for the reasons I've outlined above. In the absence of a dedicated repository, a manager seems the next best thing to me.

mvdbeek commented on lib/galaxy/managers/users.py:
```diff
@@ -840,13 +837,20 @@ def _add_parsers(self):
+    return session.scalars(stmt).all()
+
+
+# The get_user_by_email and get_user_by_username functions may be called from
```
How about a different function for the tool shed, then you also don't need to use `trans.model.User` ?

jdavcs replied:
I tried - that was my first solution. It didn't work: tool shed is tightly coupled: this function is called indirectly as part of the authentication process (as far as I remember) by both galaxy and tool shed. Changing this behavior requires refactoring tool shed code which doesn't have sufficient test coverage, and I was reluctant to touch it as part of this PR.

mvdbeek commented on lib/galaxy/managers/users.py:
```diff
@@ -840,13 +837,20 @@ def _add_parsers(self):
+# the model_class argument.
+def get_user_by_email(session, email: str, model_class=User):
+    stmt = select(model_class).filter(model_class.email == email).limit(1)
+    return session.scalars(stmt).first()
```
Is that the same as `get_user_by_username` ? It effectively looks like `.one_or_none()` ?

jdavcs replied:
`one_or_none` will raise an exception on multiple results, `first` will not. The `username` attribute is unique, but `email` is not; using `one_or_none` would thus raise an error in a potentially legitimate case, I think. I didn't want to change this behavior without looking closer at all the use cases within this PR. If you'd rather I look into this within this PR, I can do that, of course.

mvdbeek replied:
sorry, meant to write this on `get_user_by_username`. previously no result returned None, now it will blow up, so that behavior has changed.

jdavcs replied:
Sorry, why will it blow up? `first` returns first result of `None` if no result is present. [Same behavior as before](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/users.py#L843). What am I missing?

mvdbeek replied:
Ooops, didn't know that first behaves like one_or_none

mvdbeek commented on lib/galaxy/managers/workflows.py:
```diff
@@ -1972,3 +1978,9 @@ def import_workflow(self, workflow, **kwds):
         )
+
+
+def count_stored_workflow_user_assocs(session: Session, user, stored_workflow) -> int:
```
Move into user class ?

jdavcs replied:
You mean the user manager class? I don't mind, although I feel we are querying workflows by user criteria here, not users by workflow criteria. If all queries like "get this and that stuff for this user" were located in the User model class, then yes; but I was hoping we could move such methods out of models and into [managers/repositories/or somewhere else] - see my comment on History.

mvdbeek replied:
No, it's not a manager method. You can even set this up as a regular relationship if you want.

jdavcs replied:
Done.

In my opinion, just because one can setup a regular relationship for an associated model doesn't mean one should do that. You have a simple task: count the number of associations between `A` and `B`. A lightweight function that does just that and nothing more is the simplest solution here, I think.

mvdbeek commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -2177,12 +2184,12 @@ def from_json(self, value, trans, other_values=None):
                 raise ParameterValueError("more than one dataset supplied to single input dataset parameter", self.name)
             if len(rval) > 0:
-                rval = rval[0]
+                rval = rval[0]  # type:ignore[assignment]
```
```suggestion
                return rval[0]
```
probably better than ignoring the type

mvdbeek commented on lib/galaxy/webapps/base/webapp.py:
```diff
@@ -961,12 +964,13 @@ def get_most_recent_history(self):
         try:
-            recent_history = (
-                self.sa_session.query(self.app.model.History)
+            stmt = (
+                select(self.app.model.History)
                 .filter_by(user=user, deleted=False)
                 .order_by(self.app.model.History.update_time.desc())
-                .first()
+                .limit(1)
             )
+            recent_history = self.sa_session.scalars(stmt).first()
         except NoResultFound:
```
Does sqlalchemy still raise NoResultFound and should this better be `one_or_none()` ?

mvdbeek replied:
nevermind it's at least one ...

mvdbeek commented on lib/galaxy/webapps/galaxy/api/library_contents.py:
```diff
@@ -101,19 +98,7 @@ def traverse(folder):
-        except MultipleResultsFound:
-            raise exceptions.InconsistentDatabase("Multiple libraries found with the same id.")
-        except NoResultFound:
-            raise exceptions.RequestParameterInvalidException("No library found with the id provided.")
```
This still seems like a legit case and we should raise RequestParameterInvalidException ?

jdavcs replied:
The first one is not possible: The id is the pkey, so duplicate records won't exists in the database. The second one is possible; I was wondering if we really need it? We retrieve records by id everywhere and rarely check for this error. If we do want it, it would make sense to do this for all models, not just Library instances?

mvdbeek replied:
> The first one is not possible:

That's why I commented on the line I commented on, instead of a range of lines.


> We retrieve records by id everywhere and rarely check for this error.

We sure check this in many, if not all places where values come in from the API. Every missing check for incoming values is a bug, no need to make it worse here.

jdavcs replied:
Done. Sorry - I was responding from PR's main page - 4 lines were displayed as highlighted, I didn't realize you commented on the last one only.

mvdbeek commented on lib/galaxy/webapps/galaxy/controllers/history.py:
```diff
@@ -513,8 +515,13 @@ def display_by_username_and_slug(self, trans, username, slug, **kwargs):
+        user = session.scalars(stmt).first()
+
+        stmt = select(model.History).filter_by(user=user, slug=slug, deleted=False).limit(1)
+        history = session.scalars(stmt).first()
```
```suggestion
        user = session.scalars(select(model.User).filter_by(username=username).limit(1)).first()
        history = session.scalars(select(model.History).filter_by(user=user, slug=slug, deleted=False).limit(1)).first()
```
?

mvdbeek commented on lib/galaxy/webapps/galaxy/services/quotas.py:
```diff
@@ -1,17 +1,14 @@
-from galaxy.managers.quotas import QuotaManager
+from galaxy.managers.groups import get_group_by_name
+from galaxy.managers.quotas import (
+    get_quotas,
```
get_quotas is only used here, so keep it here ?

jdavcs replied:
Same as with the other services module - ideally, I think, move low-level SA code out of services? Even if it's used only there? Even if it looks trivial (and of course it is), it's still low-level SA code constructing a query against the database. I do feel it should be located separately from the business logic.. If you feel otherwise, I'll certainly change it, and we can return to this topic later, outside of the context of this PR.

mvdbeek replied:
Can you keep it here ? It's the minimum of refactoring we can agree on.

jdavcs replied:
Done.

mvdbeek reviewed (commented):
I like this a lot more, thank you @jdavcs!

jdavcs committed:
Update lib/galaxy/tools/parameters/basic.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

jdavcs committed:
Update lib/galaxy/webapps/galaxy/controllers/history.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

jdavcs committed:
Move get_quotas into services

jdavcs committed:
Raise exception if no Library found in api method

jdavcs committed:
Move get_fasta_hdas_by_history into services

jdavcs committed:
Move method into User model class

jdavcs committed:
Check if user is not null before accessing user attribute
